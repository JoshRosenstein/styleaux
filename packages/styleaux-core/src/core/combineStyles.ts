import { Style } from '@styleaux/types';
import { GetProps, PropStyleArrayFunc, Props } from './types';
type PropsOrGetProps<P, FNS extends any[]> = [P] extends [never]
  ? Partial<GetProps<FNS>>
  : Partial<P>;

/**
 * combineStyles
 * --------------
 * TODO
 *
 * ![Random](https://www.fillmurray.com/180/180)
 */
export function combineStyles<
  P extends Props = Props,
  FNS extends any[] = PropStyleArrayFunc<Props>[]
>(...fns: FNS) {
  return (props: PropsOrGetProps<P, FNS>): Style[] =>
    fns.reduce((acc, fn) => {
      const s = fn(props);
      return s ? acc.concat(s) : acc;
    }, []);
}
