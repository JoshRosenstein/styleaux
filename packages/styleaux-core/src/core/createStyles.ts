import { WithTheme, EtractNonResponsiveInputType, IStyles } from "./types";

import { mapObj, toArray, reduce } from "@roseys/futils";
import { isFunction, isPlainObject } from "typed-is";

import { STYLE_PROPS_KEY,STATIC_STYLES_KEY,DEFAULT_MEDIA_KEY} from "../constants";

import { getMedia, getThemeMedia,getDefaultMedia } from "../getters/index";

import { createWarnOnce } from "../utils/warn-once";

const warnOnce = createWarnOnce("createStyle");

function handleStyle(style, input, props, mediaKey) {
  // selector
  if (isFunction(input)) {
    return input(props, value => handleStyle(style, value, props, mediaKey));
  }

  if (isFunction(style)) {
    return style(input, props, mediaKey);
  }

  return input === true ? style : null;
}

export type WithStyleProps<S>={
  styles: S;
  [STYLE_PROPS_KEY]: EtractNonResponsiveInputType<S>;
}

export function createStyles<
  S extends {},
  Media extends {} = never,
  Theme extends {} = never,

>(
  styles: S,
  config?: { defaultTheme?: Theme; queryHandler?: Function }
): ((props: WithTheme<EtractNonResponsiveInputType<S>, Theme, Media>) => IStyles) & WithStyleProps<S>

export function createStyles<S extends {}>(
  styles: S,
  config?: { defaultTheme?: {}; queryHandler?: Function }
): (< Media extends {} = never,Theme extends {} = never>(
  props: WithTheme<EtractNonResponsiveInputType<S>, Theme, Media>
) => IStyles) & { styles: S; styleProps: EtractNonResponsiveInputType<S> };

export function createStyles<S extends {}>(styles:S) {
  function getStyles(props) {
    const media = getThemeMedia(props);
    const defaultMediaKey = getDefaultMedia(props);
const statics=styles[STATIC_STYLES_KEY]
const init= statics?[statics]:[]
    function mapStyles(input, mediaKey, style) {
      const hasMediaKey =
        Boolean(mediaKey) && mediaKey !== defaultMediaKey && mediaKey !== DEFAULT_MEDIA_KEY;

      // value with `theme.media` keys: { all: 0, M: 1 }
      if (isPlainObject(input)) {
        return mapObj(input, (value, key) => mapStyles(value, key, style));
      }
      const query = getMedia(
        mediaKey === undefined ? defaultMediaKey : mediaKey,
        media
      );

      if (hasMediaKey && !query) {
        warnOnce("Could not find mediaKey: " + mediaKey);
        return undefined;
      }
      // general prop style
      const sty = handleStyle(style, input, props, mediaKey);

      return query ? { [`@media ${query}`]: sty } : sty;
    }

    return reduce(
      (acc, propValue, propName) =>
        acc.concat(
          toArray(styles[propName]).map(
            style => mapStyles(propValue, undefined, style) || []
          )
        ),
        init,
      props
    );
  }

  return Object.assign(getStyles, {
    styles
  });
}
