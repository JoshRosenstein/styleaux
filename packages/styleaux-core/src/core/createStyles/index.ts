import { mapObj, toArray } from '@roseys/futils';
import { DEFAULT_MEDIA_KEY } from '../../constants';
import { createWarnOnce, ensureMQ } from '../../utils';
import { Style, StyleResult, Nil } from '@styleaux/types';
import { StaticOrStyleFunc, StyleInputFromProps } from './types';
import { getDefaultMedia, getMedia, getThemeMedia } from '../../getters/index';
import { isArray, isFunction, isNil, isNumeric, isPlainObject } from 'typed-is';
import {
  MediaKey,
  Props,
  PropStyleArrayFunc,
  WithTheme,
  StyleFunc,
} from '../types';
const warnOnce = createWarnOnce('createStyle');

function handleStyleReturn(
  query: string | number | Nil,
  style: Style | Nil,
  mediaKey: string | false,
): Style | Nil {
  if (mediaKey && !query) {
    warnOnce('Could not find mediaKey: ' + mediaKey);
    return null;
  }

  return isNil(style) ? null : query ? { [ensureMQ(query)]: style } : style;
}

function handleStyle(
  style: Style | StyleFunc,
  input: boolean | StyleResult | StyleFunc,
  props: unknown,
  mediaKey: MediaKey,
): Style | Nil {
  // selector
  if (isFunction(input)) {
    return input(
      props,
      (value) => handleStyle(style, value, props, mediaKey),
      mediaKey,
    );
  }

  if (isFunction(style)) {
    return style(input, props, mediaKey);
  }

  return input === true ? style : null;
}

/**
 * Create styles from {@link Object} with keys that represents component `prop` and
 * the value is a `style` that will be applied in the order it was passed.
 * @remarks
 * staticOrStyleFunc- is treated as initial style. (may change)
 *
 * ```js
 * { [prop]: style | (input, props, mediaKey) => style }
 * ```
 *
 * @param styles - {propName:( style | (input,props,key)=>style)}
 * @param [staticOrStyleFunc] - Optional param to pass in static styles or custom single arg(props) function
 *
 * @example
 *
 *  const style=createStyles({
 *    display: rule('display')
 *  }))
 *
 */

export function createStyles<
  P extends Props = Props,
  Theme extends {} = never,
  Media extends {} = never,
  PX = WithTheme<P, Theme, Media>
>(
  styles: StyleInputFromProps<P>,
  staticOrStyleFunc?: StaticOrStyleFunc<P, PX>,
): PropStyleArrayFunc<PX> {
  function getStyles(props: PX): Style[] {
    const media = getThemeMedia(props);
    const defaultMediaKey = getDefaultMedia(props);
    const mediaKeys = Object.keys(media);
    let initial: Style[] = isNil(staticOrStyleFunc)
      ? []
      : isFunction(staticOrStyleFunc)
      ? toArray(staticOrStyleFunc(props))
      : toArray(staticOrStyleFunc);

    function mapStyles(
      input: boolean | StyleResult | StyleFunc,
      style: Style | StyleFunc,
      mediaKey: string = '',
    ) {
      /// Handling responsive ObjLit
      if (isPlainObject(input)) {
        return mapObj(input, (value, key) =>
          mapStyles(value, style, key as string),
        );
      }

      /// Handling responsive Array
      if (isArray(input)) {
        return input.map((value, key) =>
          mapStyles(value, style, key.toString()),
        );
      }

      const hasMediaKey =
        mediaKey && mediaKey !== DEFAULT_MEDIA_KEY && mediaKey !== '0';

      let mKey = mediaKey;

      if (hasMediaKey) {
        if (isNumeric(mediaKey)) {
          mKey = mediaKeys[mediaKey];
        }
      }

      return handleStyleReturn(
        getMedia(mKey || defaultMediaKey, media),
        handleStyle(style, input, props, mKey),
        hasMediaKey && mediaKey,
      );
    }

    return Object.keys(styles).reduce(
      (acc, styleKey) =>
        acc.concat(
          isNil(styleKey && props[styleKey])
            ? []
            : toArray(styles[styleKey]).map(
                (style) => mapStyles(props[styleKey], style) || [],
              ),
        ),
      initial,
    );
  }

  return getStyles;
}
