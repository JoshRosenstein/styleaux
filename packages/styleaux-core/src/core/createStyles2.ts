import {ObjectInterpolation3} from '../cssTypes2'
import {mapObj, toArray} from '@roseys/futils'
import {isFunction, isPlainObject, isNil, isNumeric, isArray} from 'typed-is'

import {DEFAULT_MEDIA_KEY} from '../constants'
//import {InferResponsivePropValue} from './types'
import {getMedia, getThemeMedia, getDefaultMedia} from '../getters/index'

import {createWarnOnce} from '../utils/warn-once'

const warnOnce = createWarnOnce('createStyle')

export enum CreateStyleKeys {
  arg1='style',
  arg2='staticOrStyleFunc'

}
function handleStyle(style, input, props, mediaKey) {
  // selector
  if (isFunction(input)) {
    return input(props, value => handleStyle(style, value, props, mediaKey))
  }

  if (isFunction(style)) {
    return style(input, props, mediaKey)
  }

  return input === true ? style : null
}

type PartialRecord<K extends keyof any, T> = {[P in K]?: T}

type PartialObj<T> = PartialRecord<string,T>

type CSSObj = ObjectInterpolation3<undefined>
type StyleInputValueFunction =((...args: any[]) => CSSObj)

//type StyleInputPropsFunction<P> =((props: P) => CSSObj)

type StyleInputValue =
  | CSSObj
  | StyleInputValueFunction
  | StyleInputValueFunction[]

  type StyleInputValueP<T,P> =
  | CSSObj
  | ((input:Extract<T,boolean | string | number > ,props?:P,mediakey?:string | number,...args: any[]) => CSSObj)
  | ((input:Extract<T,boolean | string | number > ,props?:P,mediakey?:string | number,...args: any[]) => CSSObj)[]

type StyleInput = PartialObj<StyleInputValue>

//type StyleInputFromPropsCheck<P>= keyof P extends string? StyleInputFromProps<P>:Record<string,StyleInputValue>

type StyleInputFromProps<P> = {[K in keyof P]?: StyleInputValueP<P[K],P>}

//type GetPP<Style, Props, KP=keyof Props>={ [P in Extract<Style,string>]?: P extends KP? }

export type InterpolationFunc<P>=(props:P)=>CSSObj

export type InterpolationFuncArray<P> = (props: P) => CSSObj[]

 type Values<T extends object> = T[keyof T];

export interface CreateStyles2<
PROPS extends {}=any,
S extends StyleInput = StyleInputFromProps<PROPS>,
A2=CSSObj | InterpolationFunc<PROPS>  | InterpolationFuncArray<PROPS>
>{
  (
    styles: S,
    staticOrStyleFunc?: A2,
  ): InterpolationFuncArray<PROPS>

  [CreateStyleKeys.arg1]:S,
  [CreateStyleKeys.arg2]:A2,
}


export function createStyles2<
  PROPS extends {}=any,
  S extends StyleInput= StyleInputFromProps<PROPS>
>(
  styles: S,
  staticOrStyleFunc?: CSSObj | InterpolationFunc<PROPS>  | InterpolationFuncArray<PROPS>,
): InterpolationFuncArray<PROPS> {
  function getStyles(props: PROPS): CSSObj[] {
    const media = getThemeMedia(props)
    const defaultMediaKey = getDefaultMedia(props)
    const mediaKeys = Object.keys(media)
    let initial = isNil(staticOrStyleFunc)?[]: isFunction(staticOrStyleFunc)?toArray(staticOrStyleFunc(props)):toArray(staticOrStyleFunc) as CSSObj[]



    function mapStyles(input:unknown, style:Values<S>, mediaKey?: string) {
      const hasMediaKey =
        Boolean(mediaKey) && mediaKey !== DEFAULT_MEDIA_KEY && mediaKey !== '0'
      let mKey = mediaKey

      if (hasMediaKey) {
        if (isNumeric(mediaKey)) {
          mKey = mediaKeys[Number(mediaKey)]
        }
      }

      /// Only object literals can be responsive
      if (isPlainObject(input)) {
        return mapObj(input, (value, key) =>
          mapStyles(value, style, key as string),
        )
      }

      /// Handling responsive Array
      if (isArray(input)) {
        return input.map((value, key) =>
          mapStyles(value, style, key.toString()),
        )
        // mapObj(input, (value, key) =>
        //   mapStyles(value, style, key as string),
        // )
      }

      const query = getMedia(mKey || defaultMediaKey, media)

      if (hasMediaKey && !query) {
        warnOnce('Could not find mediaKey: ' + mediaKey)
        return undefined
      }
      // general prop style
      const sty = handleStyle(style, input, props, mKey)

      return query ? {[`@media ${query}`]: sty} : sty
    }

    return Object.keys(styles).reduce(
      (acc, styleKey) =>
        acc.concat(
          isNil(styleKey && props[styleKey])
            ? []
            : toArray(styles[styleKey]).map(
                style => mapStyles(props[styleKey], style) || [],
              ),
        ),
      initial,
    )
  }

  return Object.assign(getStyles, {
    [CreateStyleKeys.arg1]:styles,
    [CreateStyleKeys.arg2]:staticOrStyleFunc,
  })
}

