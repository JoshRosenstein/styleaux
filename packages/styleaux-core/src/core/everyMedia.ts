import {  identity } from '@roseys/futils';
import { isPlainObject } from 'typed-is';
import { getThemeMedia } from '../getters';
import { createWarnOnce } from '../utils/warn-once';


const warnOnce = createWarnOnce('everyMedia');

const has = (a: string[], b: string[]) => b.some(key => a.includes(key));
//const identity = <T>(v: T) => v;

export const everyMedia = (props, value, wrapper = identity) => {
  const media = getThemeMedia(props);

  if (isPlainObject(value)) {
    const mediaKeys = Object.keys(media);
    const valueKeys = Object.keys(value);

    if (has(mediaKeys, valueKeys)) {
      return (valueKeys as any[]).reduce((acc, key) => {
        if (mediaKeys.includes(key as string)) {
          const q = media[key];
          const v=wrapper(value[key])
          if(!v){return acc}
          if (q) {
            acc[`@media ${q}`] = wrapper(value[key]);
            return acc;
          } else {
            return Object.assign(acc, wrapper(value[key]));
          }
        }
        warnOnce(`Could not Find media for key %d`, key);
        return acc;
      }, {});

    }
  }

  return wrapper(value);
};
