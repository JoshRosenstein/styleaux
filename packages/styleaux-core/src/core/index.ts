export * from './createStyles';
export * from './combineStyles';
export * from './everyMedia';
export * from './style';
export * from './styler';
export * from './rule';
export * from './themeStyle';
export * from './types';
