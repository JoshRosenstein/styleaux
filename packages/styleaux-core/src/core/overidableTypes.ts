export interface Media{
  sm:never
  md:never
  lg:never
  }
