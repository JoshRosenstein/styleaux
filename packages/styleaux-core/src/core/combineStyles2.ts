

import {InferPropsFromFunctionsArgument} from './combineStyles'
import {Simplify} from '../types'
import {ObjectInterpolation3} from '../cssTypes2'

//  const compose = <P>(...funcs: Array<typeof createStyles2>) => {
//   const func = (props:P) => {
//     const n = funcs.map(fn => fn(props))
//     return n
//   }

//   return func
// }

// interface BProps{

//   b?:number
//   d?:string
// }
// interface AProps{
// a?:boolean
// }

// interface CombinedProps extends BProps,AProps{
// }

// const a=createStyles2<AProps>({'a':{margin:1}},props=>({margin:props.a?1:2}))
// const b=createStyles2<BProps>({'b':{margin:1},d:(_input,_props,_mediakey)=>({backgroundColor:'blue'})},props=>({margin:props.b?1:2}))

// const c=combineStyles2(a,b)
// const d=combineStyles2<ResponsiveObject<CombinedProps,{sm:string}>>(a,b)({a:{sm:true}})

// {
// c
// d
// }


type GetProps<P,FNS extends any[]>=[P] extends [never]? Simplify<InferPropsFromFunctionsArgument<FNS>>:P

export function combineStyles2<P extends {}= never,FNS extends any[]=([P] extends [never]? any: (p:any)=>ObjectInterpolation3[] )[] >(
  ...fns: FNS
) {


  return (props:GetProps<P,FNS>)=>fns.reduce(
    (acc,fn) => {

      return [...acc,...fn(props)]
   },[]).filter(Boolean)

  // const [styles,styleOrFunc] = fns.reduce(
  //   ([arg1,arg2], fn) => ([{
  //     ...arg1,
  //     ...fn[CreateStyleKeys.arg1],
  //   },
  //   fn[CreateStyleKeys.arg2] ? arg2.concat(fn[CreateStyleKeys.arg2]):arg2]),
  //   [{},[]] as any,
  // )

  // return createStyles2<[P] extends [never]? InferPropsFromFunctionsArgument<FNS>:P>(styles,styleOrFunc[0] && styleOrFunc)

}

// const a=combineStyles2(
//   (props:{color:string})=>({color:'red'}),
//   (props:{colorss:string})=>({color:'red'})
//   )

//   type A=Simplify<Arg1<typeof a>>
