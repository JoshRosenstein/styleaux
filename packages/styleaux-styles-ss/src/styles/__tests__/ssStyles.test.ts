import { combineStyles, WithTheme, GetProps } from '@styleaux/core';
import {
  space,
  color,
  width,
  fontSize,
  size,
  gridGap,
  gridRowGap,
  borders,
} from '../';
const theme = {
  colors: {
    blue: '#07c',
    black: '#111',
  },
};

test('returns color values from theme', () => {
  const a = color({ theme, color: 'blue', bg: 'black' });
  expect(a).toEqual([{ color: '#07c' }, { backgroundColor: '#111' }]);
});

test('returns raw color values', () => {
  const a = color({
    theme,
    color: 'inherit',
    bg: 'tomato',
  });
  expect(a).toEqual([{ color: 'inherit' }, { backgroundColor: 'tomato' }]);
});

test('backgroundColor prop overrides bg prop', () => {
  const a = color({
    backgroundColor: 'tomato',
    bg: 'blue',
  });

  expect(a).toEqual([{ backgroundColor: 'tomato' }]);
});

test('returns a pixel font-size', () => {
  const a = fontSize({ fontSize: 48 });
  expect(a).toEqual([{ fontSize: '48px' }]);
});

test('uses a default font-size scale', () => {
  const a = fontSize({
    fontSize: 2,
    theme: { fontSizes: [12, 14, 16, 20, 24, 32, 48, 64, 72] },
  });
  expect(a).toMatchSnapshot();
});

test('returns a string font-size', () => {
  const a = fontSize({ fontSize: '2em' });
  expect(a).toMatchSnapshot();
});

test('returns a percentage based width', () => {
  const a = width({ width: 1 / 2 });
  expect(a).toMatchSnapshot();
});

test('returns a pixel based width', () => {
  const a = width({ width: 256 });
  expect(a).toMatchSnapshot();
});

test('returns a string width', () => {
  const a = width({ width: 'auto' });
  expect(a).toMatchSnapshot();
});

test('returns an array of style objects', () => {
  const styles = space({
    m: '4px',
  });
  expect(styles).toMatchSnapshot();
});

test('returns 0 values', () => {
  const styles = space({ m: 0 });
  expect(styles).toMatchSnapshot();
});

test('returns negative pixel values', () => {
  const styles = space({ m: -2 });
  expect(styles).toMatchSnapshot();
});

test('returns negative em values', () => {
  const styles = space({ m: '-16em' });
  expect(styles).toMatchSnapshot();
});

test('returns negative theme values', () => {
  const styles = space({
    theme: {
      space: [0, 4, 8],
    },
    m: -2,
  });
  expect(styles).toMatchSnapshot();
});

test('returns positive theme values', () => {
  const styles = space({
    theme: {
      space: [0, '1em', '2em'],
    },
    m: 2,
  });
  expect(styles).toMatchSnapshot();
});

test('returns responsive values', () => {
  const spaceWithTheme = combineStyles<
    WithTheme<GetProps<typeof space>, { media: string[] }, string[]>
  >(space);
  const styles = spaceWithTheme({
    m: [0, 2, 3],
    theme: {
      media: [
        '@media screen and (min-width: 40em)',
        '@media screen and (min-width: 52em)',
      ],
    },
  });

  expect(styles).toMatchSnapshot();
});

test('returns aliased values', () => {
  const styles = space({
    px: 2,
  });

  expect(styles).toMatchSnapshot();
});

test('returns string values from theme', () => {
  const styles = space({
    theme: {
      space: [0, '1em'],
    },
    padding: 1,
  });
  expect(styles).toMatchSnapshot();
});

test('returns negative string values from theme', () => {
  const styles = space({
    theme: {
      space: [0, '1em'],
    },
    margin: -1,
  });
  expect(styles).toMatchSnapshot();
});

test('returns values from theme object', () => {
  const styles = space({
    theme: {
      space: { sm: 1 },
    },
    margin: 'sm',
  });

  expect(styles).toMatchSnapshot();
});

test('pl prop sets paddingLeft', () => {
  const styles = space({ pl: 2 });
  expect(styles).toMatchSnapshot();
});

test('pl prop sets paddingLeft 0', () => {
  const styles = space({ pl: 0 });
  expect(styles).toMatchSnapshot();
});

// test.skip('px prop overrides pl prop', (t) => {
//   const styles = space({
//     pl: 1,
//     px: 2,
//   });

//   // [ { paddingLeft: '2px', paddingRight: '2px' },
//   // { paddingLeft: '4px' } ]
//   expect(styles).toMatchSnapshot();
// });

test('size returns width and height', () => {
  const styles = size({
    size: 4,
  });

  expect(styles).toMatchSnapshot();
});

// grid
test('gridGap returns a scalar style', () => {
  const a = gridGap({
    theme: {
      space: [0, 2, 4, 8],
    },
    gap: 3,
  });

  expect(a).toMatchSnapshot();
});

test('gridRowGap returns a scalar style', () => {
  const a = gridRowGap({
    theme: {
      space: [0, 2, 4, 8],
    },
    rowGap: 3,
  });

  expect(a).toMatchSnapshot();
});

test('borders prop returns correct sequence', () => {
  const a = borders({
    borderBottom: '1px solid',
    borderWidth: '2px',
    borderStyle: 'dashed',
    borderColor: 'red',
  });
  expect(a).toMatchSnapshot();
});
