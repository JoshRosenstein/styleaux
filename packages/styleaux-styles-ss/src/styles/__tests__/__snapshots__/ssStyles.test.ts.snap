// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`borders prop returns correct sequence 1`] = `
Array [
  Object {
    "borderBottom": "1px solid",
  },
  Object {
    "borderWidth": "2px",
  },
  Object {
    "borderStyle": "dashed",
  },
  Object {
    "borderColor": "red",
  },
]
`;

exports[`gridGap returns a scalar style 1`] = `
Array [
  Object {
    "gap": "8px",
  },
]
`;

exports[`gridRowGap returns a scalar style 1`] = `
Array [
  Object {
    "rowGap": "8px",
  },
]
`;

exports[`pl prop sets paddingLeft 0 1`] = `
Array [
  Object {
    "paddingLeft": 0,
  },
]
`;

exports[`pl prop sets paddingLeft 1`] = `
Array [
  Object {
    "paddingLeft": "8px",
  },
]
`;

exports[`returns 0 values 1`] = `
Array [
  Object {
    "marginBottom": 0,
    "marginLeft": 0,
    "marginRight": 0,
    "marginTop": 0,
  },
]
`;

exports[`returns a percentage based width 1`] = `
Array [
  Object {
    "width": "50%",
  },
]
`;

exports[`returns a pixel based width 1`] = `
Array [
  Object {
    "width": "256px",
  },
]
`;

exports[`returns a string font-size 1`] = `
Array [
  Object {
    "fontSize": "2em",
  },
]
`;

exports[`returns a string width 1`] = `
Array [
  Object {
    "width": "auto",
  },
]
`;

exports[`returns aliased values 1`] = `
Array [
  Object {
    "paddingLeft": "8px",
    "paddingRight": "8px",
  },
]
`;

exports[`returns an array of style objects 1`] = `
Array [
  Object {
    "marginBottom": "4px",
    "marginLeft": "4px",
    "marginRight": "4px",
    "marginTop": "4px",
  },
]
`;

exports[`returns negative em values 1`] = `
Array [
  Object {
    "marginBottom": "-16em",
    "marginLeft": "-16em",
    "marginRight": "-16em",
    "marginTop": "-16em",
  },
]
`;

exports[`returns negative pixel values 1`] = `
Array [
  Object {
    "marginBottom": "-8px",
    "marginLeft": "-8px",
    "marginRight": "-8px",
    "marginTop": "-8px",
  },
]
`;

exports[`returns negative string values from theme 1`] = `
Array [
  Object {
    "marginBottom": "-1em",
    "marginLeft": "-1em",
    "marginRight": "-1em",
    "marginTop": "-1em",
  },
]
`;

exports[`returns negative theme values 1`] = `
Array [
  Object {
    "marginBottom": "-8px",
    "marginLeft": "-8px",
    "marginRight": "-8px",
    "marginTop": "-8px",
  },
]
`;

exports[`returns positive theme values 1`] = `
Array [
  Object {
    "marginBottom": "2em",
    "marginLeft": "2em",
    "marginRight": "2em",
    "marginTop": "2em",
  },
]
`;

exports[`returns responsive values 1`] = `
Array [
  Array [
    Object {
      "marginBottom": 0,
      "marginLeft": 0,
      "marginRight": 0,
      "marginTop": 0,
    },
    Object {
      "@media screen and (min-width: 40em)": Object {
        "marginBottom": "8px",
        "marginLeft": "8px",
        "marginRight": "8px",
        "marginTop": "8px",
      },
    },
    Object {
      "@media screen and (min-width: 52em)": Object {
        "marginBottom": "16px",
        "marginLeft": "16px",
        "marginRight": "16px",
        "marginTop": "16px",
      },
    },
  ],
]
`;

exports[`returns string values from theme 1`] = `
Array [
  Object {
    "paddingBottom": "1em",
    "paddingLeft": "1em",
    "paddingRight": "1em",
    "paddingTop": "1em",
  },
]
`;

exports[`returns values from theme object 1`] = `
Array [
  Object {
    "marginBottom": "1px",
    "marginLeft": "1px",
    "marginRight": "1px",
    "marginTop": "1px",
  },
]
`;

exports[`size returns width and height 1`] = `
Array [
  Object {
    "height": "4px",
    "width": "4px",
  },
]
`;

exports[`uses a default font-size scale 1`] = `
Array [
  Object {
    "fontSize": "16px",
  },
]
`;
